modules/gapi/src/api/freetype_mask_creator.hpp

```diff
@@ -0,0 +1,136 @@
+#ifdef HAVE_FREETYPE
+
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
+
+#ifndef OPENCV_FREETYPE_MASK_CREATOR_HPP
+#define OPENCV_FREETYPE_MASK_CREATOR_HPP
+
+#include "api/render_priv.hpp"
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+
+namespace cv   {
+namespace gapi {
+namespace wip  {
+namespace draw {
+
+class FreeTypeBitmaskCreator : public IBitmaskCreator
+{
+public:
+    FreeTypeBitmaskCreator()
+    {
+        FT_Init_FreeType(&m_library);
+
+        const char* font = std::getenv("FREETYPE_FONT");
+
+        if (!font) {
+            util::throw_error(std::logic_error("Please set text font to ${FREETYPE_FONT}"));
+        }
+
+        FT_New_Face(m_library, font, 0, &m_face);
+    }
+
+    virtual const cv::Size& computeMaskSize() override
+    {
+        m_slot = m_face->glyph;
+        m_mask_size = cv::Size(0, 0);
+        int bmp_w          = 0;
+        int max_bottom_pad = 0;
+        m_max_top_pad    = 0;
+
+        m_glyphs.resize(m_text.text.size());
+        m_pos.reserve(m_text.text.size());
+
+        for (size_t i = 0; i < m_text.text.size(); ++i)
+        {
+            FT_Load_Char(m_face, m_text.text[i], FT_LOAD_RENDER);
+            FT_Bitmap *bitmap = &(m_slot->bitmap);
+
+            cv::Mat(bitmap->rows, bitmap->width, CV_8UC1, bitmap->buffer, bitmap->pitch).copyTo(m_glyphs[i]);
+
+            int gl_bottom_pad = (m_slot->metrics.height - m_slot->metrics.horiBearingY) >> 6;
+            max_bottom_pad = std::max(max_bottom_pad, gl_bottom_pad);
+            m_max_top_pad = std::max(m_max_top_pad, m_slot->bitmap_top);
+
+            // FIXME why bitmap->left is negative ?
+            int gl_x_pad = m_slot->bitmap_left > 0 ? m_slot->bitmap_left + bmp_w - 1 : bmp_w;
+            auto shift = (m_slot->advance.x >> 6);
+
+
+            // FIXME why bitmap->width > shift m_slot->advance.x ?
+            if (shift < bitmap->width)
+            {
+                gl_x_pad = bmp_w;
+                bmp_w += bitmap->width;
+            }
+            else
+            {
+                bmp_w += shift;
+            }
```

Comment:
Please put a comment what the above code block does and why.

Reply:
This code computed width of text mask. Width text mask equal to sum of every glyph width.